source/core/Events.cpp

```diff
@@ -0,0 +1,750 @@
+
+    for (Int32 inputTuple = 0; inputTuple < numTuples; ++inputTuple) {
+        UInt32 eventSpecIndex = *arguments[inputTuple].eventSpecIndex;
+        if (eventSpecIndex >= esCount) {
+            THREAD_EXEC()->LogEvent(EventLog::kHardDataError, "Invalid Event Spec Index");
+            return THREAD_EXEC()->Stop();
+        }
+        TypeRef esEventDataNodeType = arguments[inputTuple].data._paramType;
+        void *esEventDataNode = arguments[inputTuple].data._pData;
+        InstructionCore *esBranchTarget = arguments[inputTuple].branchTarget;
+        if (EventMatch(eventData, eventSpecRef[eventSpecIndex], dynIndex)) {
+            // gPlatform.IO.Printf("Event match %d -> %x\n", eventSpecRef[eventSpecIndex].eventType, esBranchTarget);
+            const Int32 commonDataSize = sizeof(EventCommonData);
+            Int32 dataNodeSize = esEventDataNodeType->TopAQSize();
+            if (dataNodeSize >= commonDataSize) {
+                // TODO(spathiwa) Compute eventIndex when we support multiple events sharing event cases
+                UInt32 eventIndex = 0;
+                memcpy(esEventDataNode, &eventData, commonDataSize);
+                *EventIndexFieldPtr(esEventDataNode) = eventIndex;
+
+                if (eventData.eventDataType->TopAQSize() == dataNodeSize - commonDataSize) {
+                    // TODO(spathiwa) Should we also verify type of event data node matches event at run-time?
+                    // When we support multiple events with a shared event case, this needs to be smarter
+                    // and copy only common data.
+                    eventData.eventDataType->CopyData(eventData.pEventData, (AQBlock1*)esEventDataNode + commonDataSize);
+                }
+            } else {
+                memcpy(esEventDataNode, &eventData, dataNodeSize);
+            }
+            next = esBranchTarget;
+            break;
+        }
+    }
+    EventOracle::TheEventOracle().DoneProcessingEvent(eventQID);
+
+    return next;
+}
+
+VIREO_FUNCTION_SIGNATURE2(IsNotAUserEventRefnum, RefNumVal, Boolean)
+{
+    RefNumVal* refnumPtr = _ParamPointer(0);
+    void *userEventRef = NULL;
+    if (!refnumPtr || UserEventRefNumManager::RefNumStorage().GetRefNumData(refnumPtr->GetRefNum(), &userEventRef) != kNIError_Success)
+        _Param(1) = true;
+    else
+        _Param(1) = false;
+    return _NextInstruction();
+}
+
+VIREO_FUNCTION_SIGNATURE3(IsEQRefnum, RefNumVal, RefNumVal, Boolean);
+VIREO_FUNCTION_SIGNATURE3(IsNERefnum, RefNumVal, RefNumVal, Boolean);
+
+//------------------------------------------------------------
+DEFINE_VIREO_BEGIN(Events)
+
+    // User Events
+    DEFINE_VIREO_TYPE(UserEventRefNum, "refnum($0)")
+    DEFINE_VIREO_FUNCTION_CUSTOM(CreateUserEvent, UserEventRef_Create, "p(o(UserEventRefNum ue) io(ErrorCluster err))")
+    DEFINE_VIREO_FUNCTION_CUSTOM(GenerateUserEvent, UserEventRef_Generate, "p(io(UserEventRefNum ue) i(* element) i(Boolean highprio) io(ErrorCluster err))")
```

Comment:
Good catch, yes, for Generate and Unregister they should be inputs only.
RegisterForEvents is correct.  It creates the refnum if it comes in as not-a-ref, and uses/modifies if otherwise.